pkg/utils/release/controller.go

```diff
@@ -424,3 +450,112 @@ func (s *SuiteController) CreateReleasePipelineRoleBindingForServiceAccount(name
 	}
 	return roleBinding, nil
 }
+
+func (s *SuiteController) WaitForReleasePipelineToBeFinished(managedNamespace, devNmaespace string, releaseName string, component *appstudioApi.Component, maxRetries int) error {
+	attempts := 1
+	var pr *v1beta1.PipelineRun
+	var prList *v1beta1.PipelineRunList
+
+	for {
+		err := wait.PollImmediate(20*time.Second, 30*time.Minute, func() (done bool, err error) {
+
+			if len(releaseName) > 0 {
+				prList, err = s.GetPipelineRunListInNamespace(managedNamespace, releaseName, devNmaespace)
+				if err != nil {
+					g.GinkgoWriter.Println("PipelineRun has not been created yet.")
+					return false, nil
+				}
+			} else {
+				prList, err = s.GetAllReleasePipelineRunInNamespace(managedNamespace, devNmaespace, component)
+				if err != nil {
+					g.GinkgoWriter.Println("PipelineRun has not been created yet.")
+					return false, nil
+				}
+
+			}
+
+			pr := &prList.Items[0]
+			g.GinkgoWriter.Printf("PipelineRun %s reason: %s\n", pr.Name, pr.GetStatusCondition().GetCondition(apis.ConditionSucceeded).GetReason())
+
+			if !pr.IsDone() {
+				return false, nil
+			}
+
+			if pr.GetStatusCondition().GetCondition(apis.ConditionSucceeded).IsTrue() {
+				return true, nil
+			} else {
+				var prLogs string
+				if err = tekton.StorePipelineRun(pr, s.KubeRest(), s.KubeInterface()); err != nil {
+					g.GinkgoWriter.Printf("failed to store PipelineRun %s:%s: %s\n", pr.GetNamespace(), pr.GetName(), err.Error())
+				}
+				if prLogs, err = tekton.GetFailedPipelineRunLogs(s.KubeRest(), s.KubeInterface(), pr); err != nil {
+					g.GinkgoWriter.Printf("failed to get logs for PipelineRun %s:%s: %s\n", pr.GetNamespace(), pr.GetName(), err.Error())
+				}
+				return false, fmt.Errorf(prLogs)
+			}
+		})
+
+		if err != nil {
+			g.GinkgoWriter.Printf("attempt %d/%d: PipelineRun %q failed: %+v", attempts, maxRetries+1, pr.GetName(), err)
+			attempts++
+		} else {
+			break
+		}
+	}
+
+	return nil
+
+}
+
+// GetAllReleasePipelineRunInNamespace returns the release pipelineRuns  in namespace.
+func (s *SuiteController) GetAllReleasePipelineRunInNamespace(managedNamespace, devNamespace string, component *appstudioApi.Component) (*v1beta1.PipelineRunList, error) {
```

Comment:
1. the method name doesn't describe correctly what it's actually doing.

2. IMO in general the method should be simple, ideally doing just one thing, unless there are some workarounds needed.
I think the part for getting the snapshot should be separated, i.e. called from the spec from the test file rather than from release controller, otherwise it could lead to confusions when the method returns error like the one below: 
`error could not find snapshot in %s namespace` - why it's complaining about snapshots when I wanted to list Release PipelineRuns?

Reply:
totally agree for both points here , 

`1. the method name doesn't describe correctly what it's actually doing.`

I've changed to name **GetAllReleasePipelineRunMatchComponent** 

```
IMO in general the method should be simple, ideally doing just one thing, unless there are some workarounds needed.
I think the part for getting the snapshot should be separated, i.e. called from the spec from the test file rather than from release controller, otherwise it could lead to confusions when the method returns error like the one below:
error could not find snapshot in %s namespace - why it's complaining about snapshots when I wanted to list Release PipelineRuns?
```
I've made it out of three generic methods .